File: medium/0230_kth_smallest_element_in_a_bst/solution.py
from typing import Optional

from utils.python.nodes import TreeNode


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        curr = root
        stack = []

        values = []

        while curr or stack:
            while curr:
                stack.append(curr)
                curr = curr.left

            node = stack.pop()
            values.append(node.val)
            curr = node.right

        return values[k - 1]

File: medium/0230_kth_smallest_element_in_a_bst/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <stack>
#include <vector>

class Solution {
public:
  int kthSmallest(utils::TreeNode *root, int k) {
    utils::TreeNode *curr = root;
    std::stack<utils::TreeNode *> s;

    std::vector<int> values;

    while (curr || !s.empty()) {
      while (curr) {
        s.push(curr);
        curr = curr->left;
      }

      utils::TreeNode *node = s.top();
      s.pop();

      values.push_back(node->val);
      curr = node->right;
    }

    return values[k - 1];
  }
};
